Custom Operators for arbitrary processes

GabrielePalazzo opened the pull request:
Added the possibility to generate C++/CUDA Custom Operators for the evaluation of polarized Matrix Elements, in order to reduce execution times.

The Custom Op can be generated and compiled during runtime using `--custom_op`, i.e.:
```
madflow --custom_op
```

In order to reduce execution times, the Op isn't generated if the argument isn't used.

`custom_op_generator.py` located inside `python_package/madflow/` generates the Op for arbitrary processes. The Op is then compiled using the `makefile`.

Current issues:
 - `makefile` isn't installed and thus `madflow` cannot compile the Op if executed from the install directory;
 - the path to `nvcc` and CUDA can't be determined a priori. Currently the path defined in the makefile is correct for my laptop, however it is guaranteed to fail on other systems. I think we can safely retrieve `nvcc` with (`which nvcc`), but I have no idea regarding the path of CUDA (maybe from `$PATH`?).

I didn't apply any changes to `examples`.

Let me know if you have any questions.
Gabriele Palazzo

scarlehoff commented:
Hi @GabrielePalazzo thank you very much for this :)
Sadly I have no time this week to go through it in detail but just a few general comment, also about the issues you describe.

The changes to `madflow_exec` look perfect. 
The `makefile` as well, I would remove the things that are commented out and instead add comments explaining some of the non obvious stuff (although, for the makefile read below)
The `custom_op_operator.py` instead is gigantic. I think it would be beneficial to break it into a few submodules with the `write_` functions, the `classes` and so on separated. It would be great to create the `cuda` module I mention also below with everything a bit more organized.

I would also remove the comments (since they are now commited here, git is already tracking whatever was already written before) and write docstrings and a bit of documentation instead.

> makefile isn't installed and thus madflow cannot compile the Op if executed from the install directory;

I think the best course of action is to perform (everything) in the user directory (i.e., where the user creates the matrix elements and all of that). For that the `makefile` needs to be generated there on the fly (the lowest effort solution is to create a submodule `cuda` where one file is for instance `makefile_template.py` with a function that writes down the `makefile`). That `cuda` (python)  submodule will be automatically installed and so you will have access to the makefile upon installation.

> the path to nvcc and CUDA can't be determined a priori. Currently the path defined in the makefile is correct for my laptop, however it is guaranteed to fail on other systems. I think we can safely retrieve nvcc with (which nvcc), but I have no idea regarding the path of CUDA (maybe from $PATH?).

I would just do NVCC=nvcc, if the user has it in its path good if not, the user has to do NVCC=nvcc but it's "their" problem. I think `which nvcc` only works if the user already has it in their path anyway.

Similarly for cuda, the user should have everything in their lib/include path already if not they can `export CUDA_PATH=/what/ever/` with the correct thing in case it is non standard.

Then that gets added to the documentation and done :P 

Last but not least, I don't know whether you know this tool: https://pypi.org/project/black/ it basically stylizes python code for you, so you can do `black -l100 myfile.py ` and it will rewrite the file with an unified style. It's what we are using in all the N3PDF projects so that they look more "professional" :P (the `-l100` part is because the default a few years ago was 80 characters per line, maybe they changed it now but it feels to small for me).

GabrielePalazzo commented:
I removed the `makefile` and instead created `makefile_template.py` which contains a function (`write_makefile`) that takes as argument the directory of ME files.
I commented all flags for OpenMP, as we're currently using ThreadPool for CPU parallelization.

Inside `write_makefile` there are only function calls, as I think with this approach it is very clear what is written into the `makefile`.
Let me know if you have any suggestions or if you prefer a different approach.

> I would just do NVCC=nvcc, if the user has it in its path good if not, the user has to do NVCC=nvcc but it's "their" problem. I think which nvcc only works if the user already has it in their path anyway.
>
> Similarly for cuda, the user should have everything in their lib/include path already if not they can export CUDA_PATH=/what/ever/ with the correct thing in case it is non standard.

Here is a brief summary:
 - using `which nvcc` the compiler must be within `$PATH`. However if `nvcc` isn't found, it reverts to the C++ compiler and compiles the Op just for CPU operations (instead of returning an error).
 - The path of CUDA library files is guessed from `$PATH`. However it can be explicitly defined in two ways:
   1. at the beginning of `makefile_template.py`;
   2. as an environment variable (`export CUDA_PATH=...`).

   In any case (ii) has higher priority over (i).
 - Both C++ compiler and C++ version are defined at the beginning of `makefile_template.py` thus making it convenient for different requirements in the future.

> The custom_op_operator.py instead is gigantic. I think it would be beneficial to break it into a few submodules with the write_ functions, the classes and so on separated. It would be great to create the cuda module I mention also below with everything a bit more organized.

I agree. I'll keep you updated on that one.

> Last but not least, I don't know whether you know this tool: https://pypi.org/project/black/ it basically stylizes python code for you, so you can do black -l100 myfile.py and it will rewrite the file with an unified style. It's what we are using in all the N3PDF projects so that they look more "professional" :P (the -l100 part is because the default a few years ago was 80 characters per line, maybe they changed it now but it feels to small for me).

Done for `madflow_exec.py` and for `makefile_template.py`.

GabrielePalazzo commented:
Quick update. I split `custom_op_generator.py` into multiple modules (all starting with `op_`):
 - `op_aux_functions.py` contains functions used across the generator in addition to the sign functions used in the Custom Operator;
 - `op_classes.py` contains all classes used in the generator;
 - `op_constants.py` contains strings used by jinja for the actual generation of C++/CUDA source code;
 - `op_global_constants.py` contains all libraries and data types used by the Op. They are separated form other constants, in order to make them more accessible if necessary (i.e. different number of threads per block / different type for complex numbers / different TensorFlow libraries);
 - `op_generator.py` contains functions for creating the modified TensorFlow Python class, in addition to function for refinements of the Op (i.e. parallelization and memory optimizations);
 - `op_parser` transforms python code into smaller pieces, making them easier to understand by the transpiler;
 - `op_read.py` reads TensorFlow functions;
 - `op_syntax.py` controls the syntax;
 - `op_transpiler.py` converts the syntax;
 - `op_write_templates.py` writes jinja templates to a string (which is then saved to file).

`custom_op_generator.py` can still be made smaller: `translate(...)` contains redundancies.

Some modules require code structure changes.

scarlehoff commented:
Thanks! Let me know when this is ready for review. From tomorrow I should have a bit more time.

One quick comment, it would be good to add docstrings to the functions to understand more easily what they do, like for instance:
```python
def write_shell_name(text):
    """Adds a line to the makefile for the kernel name"""
    text += "UNAME_S := $(shell uname -s)\n"
    text = write_nl(text)
    return text
```
since even if some might seem obvious right now, after some time people forget what things did (I know I do forget).

This is in general a good thing to do, personally I even try sometimes to write the docstring before writing the function because it forces me to think "why do I need this function? What do I want to have inside? What should be the input?" it's a bit like a todo list for myself.

Since we are at the `makefile_template.py`, I don't think a function should modify its input unless strictly necessary.
For instance, you are creating a function that does:

```python
def fun(text):
    text += "A"
    return text
```

This kind of function is dangerous because these two calls are both valid but will produce very different results:

```python
initial_text = "prova"
initial_text = fun(initial_text)
```
Returns `provaA`


```python
initial_text = "prova"
initial_text += fun(initial_text)
```
Returns `provaprovaA` instead!

So you should decide either to change things in-place, so, without a return value

```python
def fun(text):
    text += "A"
```

or to return just the "inside" text

```python
def fun():
    return "A"
# and then it is called like
initial_text += fun()
```

You might need to use the text, and that might be more tricky, but in not performance-critical situations I think it would be better to do:
```python
def fun(text):
   new_text = text + "A"
   return new_text
```

GabrielePalazzo commented:
Thanks for the comment.

> or to return just the "inside" text
> ```
> def fun():
>     return "A"
> # and then it is called like
> initial_text += fun()
> ```

I modified `makefile_template.py` with this approach, and I also added docstrings to it. I'll do the same for the other files too.

The flags for OpenMP are commented out for two reasons:
 1. we're currently using ThreadPool for CPU parallelization;
 2. the Op runs on all threads with OpenMP event without those flags.

> Thanks! Let me know when this is ready for review. From tomorrow I should have a bit more time.

If you want to start to have a look at the code, feel free to start a review. The only file that needs modifications is `op_transpiler.py`.

These modules are ready:
 - `custom_op_generator.py`
 - `makefile_template.py`
 - `op_classes.py`
 - `op_constants.py`
 - `op_global_constants.py`
 - `op_generation.py` (no docstrings yet)
 - `op_parser.py` (no docstrings yet)
 - `op_write_templates.py`

Let me know if you have questions / suggestions.

GabrielePalazzo commented:
I've added comments for every function, with a brief description of arguments and return values.

The only modules that seem to big are
 - `op_syntax.py`
 - `op_transpiler.py`

GabrielePalazzo commented:
@scarlehoff , I think it's ready for review. Let me know if you have any questions.

scarlehoff commented on python_package/madflow/custom_op_generator.py:
```diff
@@ -0,0 +1,165 @@
+        matrix_source = destination + "matrix_1_" + process_name + ".py"
+        process_source = destination + "aloha_1_" + process_name + ".py"
+
+        _file_ = process_source
```
I would not change the value of `_file_` here, just use `process_source` when you need it.

GabrielePalazzo replied:
I commented out line 74 (to be removed in the next commit) and used `process_source`.

scarlehoff commented on python_package/madflow/custom_op_generator.py:
```diff
@@ -0,0 +1,165 @@
+        # --------------------------------------------------------------------------------------
+
+
+def compile(destination):
```
As said before, using `compile` as the name for a function clashes with a built-in and thus can create unexpected problems.

GabrielePalazzo replied:
I renamed it to `compile_op`.

scarlehoff commented on python_package/madflow/custom_op_generator.py:
```diff
@@ -0,0 +1,165 @@
+def compile(destination):
+    """Compiles the Custom Operator
+    destination: directory of madflow output"""
+    subprocess.check_output(["/bin/sh", "-c", "cd " + destination + "; make"])
```
```suggestion
    subprocess.run("make", cwd=destination, check=True)
```
since you are not using the output this should work just the same. If not, you might need to add `shell=True` See [docs](https://docs.python.org/3/library/subprocess.html#subprocess.run)

GabrielePalazzo replied:
Done.

scarlehoff commented on python_package/madflow/custom_op_generator.py:
```diff
@@ -0,0 +1,165 @@
+
+        signatures = []
+        for s in signatures_:
+            signatures.append(s)
```
I'm very confused about this loop, why is it not possible to do just `signatures=signatures_` or just use `signatures_` directly?

GabrielePalazzo replied:
The code after the assignment changes values inside `function_list` and `headers`, as CUDA functions are slightly different from C++ (CPU) functions.
Using the direct assignment, the assigned list is not a new copy of the previous list.

scarlehoff replied:
But when you do below

```
        signatures, signature_variables = read_signatures(signatures, signature_variables, _file_)
```

the `signatures, signature_variables` should already be copied, and if you are mutating the values inside `read_signature` then they should not be returned.

Personally, I think that the cleanest solution -in general- is always to not mutate the input so that inside `read_signature` if you need you do `signatures = copy(signatures)`

scarlehoff commented on python_package/madflow/custom_op_generator.py:
```diff
@@ -0,0 +1,165 @@
+from madflow.op_write_templates import *
+from madflow.op_syntax import *
+from madflow.op_read import *
+from madflow.op_generation import *
```
You should never import everything with a module with * it's way better to do just

`import madflow.op_write_templates as op_wt` (or whatever) and then use `op_wt.function_from_module`

otherwise if you by mistake put in `op_generation` a variable with the same name as something in `op_write_templates` everything will fail in a very non obvious way.

GabrielePalazzo replied:
Done.

scarlehoff commented on python_package/madflow/makefile_template.py:
```diff
@@ -0,0 +1,185 @@
+makefileName = "makefile"
+cppCompiler = "g++"
+cppVersion = "c++14"
+cudaPath = ""
```
```suggestion
MAKEFILE = "makefile"
CPPCOMPILER = "g++"
CPPVERSION = "c++14"
```
As a convention constant variables are usually uppercase.

scarlehoff commented on python_package/madflow/makefile_template.py:
```diff
@@ -0,0 +1,185 @@
+    return text
+
+
+def write_omp_flags():
+    """Adds flags for OpenMP parallelization"""
+    text = "ifeq ($(UNAME_S), Darwin)\n"
+    text += "OMP_CFLAGS = -Xpreprocessor -fopenmp -lomp\n"
+    text += "else\n"
+    text += "OMP_CFLAGS = -fopenmp\n"
+    text += "endif\n"
+    text += "\n"
+    return text
```
This is a suggestion that if you want you can apply to this entire module or ignore completely. One way you can take advantage of the fact that you are using functions to generate the `makefile` is to add some flexibility to them, for instance:

```suggestion
def write_omp_flags(openmp_flag="-fopenmp"):
    """Adds flags for OpenMP parallelization"""
    text = f"""ifeq ($(UNAME_S), Darwin)
OMP_CFLAGS = -Xpreprocessor {openmp_flag} -lomp
else
OMP_CFLAGS = {openmp_flag}
endif
"""
    return text
```

For isntance, at the moment your code works with gcc but it would fail with the intel compiler (for which the flag is `-qopenmp`). Now, you are not using the intel compiler so you can't even test, but by writting the function in this more flexible manner it allows for easier future extensions.

I've used the `"""` `""""` syntax and the f-string instead of adding the `\n` because I feel it was cleaner here. It's not necessary always the case of course.

GabrielePalazzo replied:
Done. I also used more f-strings in `makefile_template.py`.

scarlehoff commented on python_package/madflow/makefile_template.py:
```diff
@@ -0,0 +1,185 @@
+    makefileContent += write_commands()
+
+    # write the makefile
+    with open(destination + makefileName, "w") as fh:
+        fh.write(makefileContent)
```
If you are already using paths you can do:

```suggestion
    (destination / makefile_name).write_text(makefile_content)
```

I would also appreciate the change from camelcase to underscore, so `makefileContent` -> `makefile_content` to follow the same style as the rest of the code ([pep8](https://www.python.org/dev/peps/pep-0008/)) but that's something that I can do myself automatically after the PR is merged.

GabrielePalazzo replied:
Done. Let me know if you still find camelcase names.

scarlehoff commented on python_package/madflow/op_classes.py:
```diff
@@ -0,0 +1,75 @@
+        self.template = f_template
+
+
+class custom_operator:
```
Instead, the convention for classes is big camelcase :P 

```suggestion
class CustomOperator:
```

also, you can use for these ones python dataclasses: https://docs.python.org/3/library/dataclasses.html

```python
from dataclasses import dataclass

@dataclass
class CustomOperator:
    """docs"""
    name: str
    scope: str
    functor_name: str
```

And python will generate automatically the `__init__` as well as other goodies (such as nice printing, equality methods, etc).

I would make all functions in this module as dataclasses. When you need to act in some arguments you can add a `__post_init__` method https://docs.python.org/3/library/dataclasses.html#post-init-processing

GabrielePalazzo replied:
Done.

scarlehoff commented on python_package/madflow/op_constants.py:
```diff
@@ -0,0 +1,248 @@
+from madflow.op_global_constants import doubleType, complexType
```
```suggestion
"""Jinja templates for the generation of custom operators"""

from madflow.op_global_constants import doubleType, complexType
```

GabrielePalazzo replied:
Done.

scarlehoff commented on python_package/madflow/op_generation.py:
```diff
@@ -0,0 +1,335 @@
+from madflow.op_aux_functions import *
+
+n_events = argument("nevents", "const int", 0, False, [])
+forLoopString = "for (int it = 0; it < " + n_events.name + "; it += 1) {"
```
Why are these two lines separated (and more importantly executed) at import time?

GabrielePalazzo replied:
The number of events is not explicitly defined in TensorFlow, as it's implicitly obtained from tensor shapes. As it is used in multiple functions I thought it was better (and less error-prone) to separate these two lines.

scarlehoff reviewed (changes requested):
Thank you again for all the work.

I've already said it but the performance if fantastic (I will be running more benchmarks as I continue with the review)

| run             | Tensorflow (s)   | Cuda (s)      | Result (fb)  |
| ----------------| ---------------- | --------------|--------------|
| `g g > t t~`    | 2.22 (46.32)     | 0.64 (44.71)  | 377.3 +- 0.1 |
| `g g > t t~ g`  | 8.86 (127.7)     | 1.18 (50.81)  | 126.6 +- 0.2 |
| `p p > t t~ g`  | 13.1 (176.1)     | 1.44 (59.73)  | 134.6 +- 0.2 |
| `g g > t t~ g g`| 661.0 (66761)    | 8.82 (149.8)  | 41.6  +- 1.8 |


I've left a number of comments for the modules that I've already "parsed". They are mostly comments about style / conventions or the possibility of using python features to simplify or make the code more flexible. They of course apply to the rest of the code (such as the global imports with `*` which I'm very much against).

The logic itself of what I've read for the moment is perfectly good, I'll continue reading through the rest next week.

GabrielePalazzo commented:
Hi, I'm back after covid. I've done a few changes. The next things I'll look into are:
 - constant variables
 - paths

GabrielePalazzo commented on python_package/madflow/custom_op_generator.py:
```diff
@@ -0,0 +1,165 @@
+        .split("\n")[:-1]
+    )
+
+    for _file_ in files_list:
```
Renamed `_file_` to `subprocess_file`.

GabrielePalazzo commented on python_package/madflow/scripts/madflow_exec.py:
```diff
@@ -299,6 +304,9 @@ def madflow_main(args=None, quick_return=False):
     _generate_madgraph_process(args.madgraph_process, output_path)
+    if args.custom_op:
+        translate(str(output_path))
+        compile(str(output_path))
```
Done.

GabrielePalazzo commented on python_package/madflow/custom_op_generator.py:
```diff
@@ -0,0 +1,165 @@
+    destination: directory of madflow output"""
+
+    if destination[-1] != "/":  # Avoid weird behaviours if destination does not end with '/'
+        destination += "/"
```
Done.

GabrielePalazzo commented on python_package/madflow/custom_op_generator.py:
```diff
@@ -0,0 +1,165 @@
+        subprocess.check_output(["/bin/sh", "-c", "ls " + destination + " | grep matrix_1_"])
+        .decode("utf-8")
+        .split("\n")[:-1]
+    )
```
Done. I also removed explicit `.py` extensions, as you suggested.

scarlehoff replied:
Looks much better now, thanks! You can remove the old code (instead of leaving it commented). Git will keep a history of everything anyway.

GabrielePalazzo replied:
Ok.

scarlehoff commented:
> Hi, I'm back after covid. I've done a few changes. 

Ouch, sorry about that! I hope it wasn't too bad!

scarlehoff commented:
> Done. Should I also change to uppercase constant variables in op_global_constants.py?

Yes, the convention of having uppercase constants is very useful because allows you to quickly identify what is known not to change anywhere in the code.

GabrielePalazzo commented on python_package/madflow/makefile_template.py:
```diff
@@ -0,0 +1,199 @@
+\trm -f $(TARGETS) $(OBJECT_SRCS_CUDA)
+\trm -f gpu/*
+
+"""
```
Done.

GabrielePalazzo commented on python_package/madflow/custom_op_generator.py:
```diff
@@ -0,0 +1,173 @@
+        headers = []
+        for h in headers_:
+            headers.append(h)
+        # """
```
Done with `copy`.

GabrielePalazzo commented on python_package/madflow/custom_op_generator.py:
```diff
@@ -0,0 +1,173 @@
+        for i in range(len(function_list)):
+            function_list = check_lines(i, function_list)
+        for i in range(len(function_list)):
+            function_list = check_variables(i, function_list)
```
The first `check_variables` tries to determine the size (and type) of arrays statically defined inside functions. i.e.:
`double v;` becomes `T v[4];`

After having determined all sizes, the code can resolve all `tf.concat` left (done in `check_lines`, which also converts remaining python syntax). i.e.:
`fi  = tf.concat([v0, v1, v], axis=1)` becomes
```
    fi[0] = v0;
    fi[1] = v1;
    for (int it1 = 0; it1 < 4; it1++) {
        fi[2 + it1] = v[it1];
    }
```

The final `check_variables` ensures function arguments have correct type (and converts incorrect types).

scarlehoff replied:
I'm still confused. Why the first check variables didn't check that the arguments had the correct type (and converted the incorrect ones?)

GabrielePalazzo replied:
Fixed. Now it should work even if it's called once.
(the issue was caused a design flaw)

GabrielePalazzo commented:
> the signatures, signature_variables should already be copied, and if you are mutating the values inside read_signature then they should not be returned.
>
> Personally, I think that the cleanest solution -in general- is always to not mutate the input so that inside read_signature if you need you do signatures = copy(signatures)

I removed returns for these functions.

scarlehoff commented on python_package/madflow/op_parser.py:
```diff
@@ -0,0 +1,115 @@
+    """Read the signature from text and update signature_variables
+    signatures: list of text Signature objects
+    signature_variables: list of tf.function signatures"""
+    for i in range(len(signature_variables)):
+        for v in signature_variables[i].signature_name_list:
+            for s in signatures:
+                if s.name == v:
+                    signature_variables[i].signature_list.append(s)
```
```suggestion
    for sv in signature_variables:
        for s in signatures:
            if s.name in sv.signature_name_list:
                sv.signature_list.append(s)
```

Please check that this does the correct thing. I've removed the range/len and simplify a bit the append.

GabrielePalazzo replied:
I get a "list index out of range" error.
I think `sv` is a temporary copy. For example, if I run this code
```
elements = [1, 2, 3, 4, 5, 6, 7]
for el in elements:
    el = 2
print(elements)
```
I get `[1, 2, 3, 4, 5, 6, 7]`.

scarlehoff replied:
Where do you get the `list index out of range`. It's likely that I mistook the order or some of the variables, but I cannot see the mistake.

In your code `el` is just a pointer to something else. If you run this instead:

```python
elements = [[1], [2], [3]]
for el in elements:
    el.append("a")
print(elements)
```

you will get `[[1, 'a'], [2, 'a'], [3, 'a']]` which is similar to what you have.

GabrielePalazzo replied:
I think the error comes from `if s.name in sv.signature_name_list`. Let's take `sign` as an example.
It's signature is `[sscalar]*2`. The code looks for known signatures and converts `sscalar` into an array of signatures containing something more "detailed" (variable type and size).
Your code finds only one occurrence, effectively converting `[sscalar]*2` into a single `sscalar`.
```
sign_signature:
[Signature(name='sscalar', type='double', size=0, tensor=False, slice=[])]
```
instead of
```
sign_signature:
[Signature(name='sscalar', type='double', size=0, tensor=False, slice=[]), Signature(name='sscalar', type='double', size=0, tensor=False, slice=[])]
```


This version (without the C-style iterator) seems to work correctly:
```
for sv in signature_variables:
    for v in sv.signature_name_list:
        for s in signatures:
            if s.name == v:
                sv.signature_list.append(s)
```

scarlehoff replied:
Ah! I see. I though there could be only one occurrence. It makes sense, yes.

scarlehoff commented:
Thank you very much for everything :)

One last thing, could you move these functions to a subfolder? Something like `madflow/custom_op` (and then they get imported from there of course). If you want you can remove also the `op_` at the beginning so that things are `madflow/custom_op/parser` which I believe is very explicit in what it does.

I don't know if you know of `sed` but if you don't, this command will change all the imports automagically:

```bash
~$ sed -i 's/madflow.op_/madflow.custom_op./' <file_to_be_changed.py>
```

GabrielePalazzo commented:
Hi, in this weekend I made a few strings inside `transpiler.py` more compact. I also found a typo in the docs.

scarlehoff commented:
Thanks. I think this can already be merged.

The only issue is the famous `NaN` that we were seeing every now and then but I guess that's something to be investigated in the future.

Thank you very much :)
